File: TIA/tia_final_project/src/com/crdzbird/dao/CityDao.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.crdzbird.dao;

/**
 *
 * @author crdzbird
 */

import com.crdzbird.models.City;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class CityDao {
    
    private static final List<City> CITIES = new ArrayList<>();

    public static City create(City c) {
        Objects.requireNonNull(c, "City object cannot be null");
        if (!CITIES.contains(c)) {
            CITIES.add(c);
        }
        return c;
    }
    
    public static City remove(City c) {
        CITIES.remove(c);
        return c;
    }

    public static City get(String name) {
        for (City city : CITIES) {
            if (city.getName().equals(name)) {
                return city;
            }
        }
        return null;
    }
    
    public static City get(int index) {
        return CITIES.get(index);
    }
    
    public static int getIndex(City c) {
        return CITIES.indexOf(c);
    }

    public static List<City> getAll() {
        return new ArrayList<>(CITIES);  // Return a copy for safety
    }
}

File: TIA/tia_genetic_memetic/lib/service/city_dao.dart
import 'package:tia_genetic_memetic/models/city.dart';

class CityDao {
  static final List<City> _cities = List.empty(growable: true);

  static City create(City? c) {
    if (c == null) {
      throw ArgumentError('City object cannot be null');
    }
    if (!_cities.contains(c)) {
      _cities.add(c);
    }
    return c;
  }

  static City remove(City c) {
    _cities.remove(c);
    return c;
  }

  static City? get(String name) {
    for (City city in _cities) {
      if (city.name == name) {
        return city;
      }
    }
    return null;
  }

  static City? getAt(int index) {
    if (index < 0 || index >= _cities.length) {
      return null; // or throw an exception
    }
    return _cities[index];
  }

  static int getIndex(City c) {
    return _cities.indexOf(c);
  }

  static List<City> getAll() {
    return List<City>.unmodifiable(_cities);
  }
}
